File: langs/ocaml-ddd/navigator.ml
open Robot
open Grid
open Scents

let execute (module R: REPO) (g: grid) (r: robot) (instr: string) =
  String.iter (fun c ->
    if not r.lost then match c with
    | 'L' -> turn_left r
    | 'R' -> turn_right r
    | 'F' -> let next = next_forward r in
             if not (contains g next) then (
               if not (R.has r.position r.orientation) then
                 (R.add r.position r.orientation; r.lost <- true)
               else ()
             ) else r.position <- next
    | _ -> ()
  ) instr

File: rust-martian-ddd/martian-robot-ddd/src/domain/navigator.rs
use super::{grid::Grid, robot::Robot, scents::ScentRepository};

pub fn execute<R: ScentRepository>(
    repo: &mut R,
    grid: &Grid,
    robot: &mut Robot,
    instructions: &str,
) {
    for ch in instructions.chars() {
        if robot.lost {
            break;
        }
        match ch {
            'L' => robot.turn_left(),
            'R' => robot.turn_right(),
            'F' => {
                let next = robot.next_forward();
                if !grid.contains(next) {
                    if !repo.has_scent(robot.position, robot.orientation) {
                        repo.add_scent(robot.position, robot.orientation);
                        robot.lost = true;
                    }
                } else {
                    robot.position = next;
                }
            }
            _ => {}
        }
    }
}
